Fix for the reStructuredText preview in vscode

Milacus opened the pull request:
Added cleansing of an environment variable when calling the drawio application

- The variable is automatically set when run in certain vscode extension contexts
- The environment variable is evaluated by electron and prevents the normal application entry point

This fix is necessary to utilize the reStructuredText (sphinx) preview, directly in vscode

Modelmat commented on sphinxcontrib/drawio/__init__.py:
```diff
@@ -250,6 +250,10 @@ def _drawio_export(self, input_abspath, options, out_filename):
         if builder.config._display:
             new_env["DISPLAY"] = ":{}".format(builder.config._display)
 
+        if ('ELECTRON_RUN_AS_NODE' in new_env):
```
Please do not use braces in the if-statement and use double quotes. Running `black` locally should pick this up.

Modelmat commented on sphinxcontrib/drawio/__init__.py:
```diff
@@ -250,6 +250,10 @@ def _drawio_export(self, input_abspath, options, out_filename):
             new_env["DISPLAY"] = ":{}".format(builder.config._display)
 
+        if ('ELECTRON_RUN_AS_NODE' in new_env):
+            # This env var will cause the drawio desktop app to not run properly
```
```suggestion
            # This environment variable prevents the drawio application from starting.
            # This is automatically set within certain Visual Studio Code contexts, 
            # such as for the reStructuredText (sphinx) preview.
```

Modelmat commented on sphinxcontrib/drawio/__init__.py:
```diff
@@ -250,6 +250,10 @@ def _drawio_export(self, input_abspath, options, out_filename):
 
+        if ('ELECTRON_RUN_AS_NODE' in new_env):
+            # This env var will cause the drawio desktop app to not run properly
+            new_env.pop("ELECTRON_RUN_AS_NODE")
```
Perhaps consider not using an `if` and instead using `new_env.pop("ELECTRON_RUN_AS_NODE", None)`.

Milacus commented:
Resolved the review findings as suggested

Modelmat commented:
Hi

This seems fine.
For some reasons the tests seem to be broken - they seem to be fine on master, so not sure what's up with that.

I'm currently in final exam period (well, whenever it happens after COVID dies down here again) so I'm busy studying and don't have time to try figure that out, sorry. 
If you'd like to have a look and see what's going wrong I can merge a fix but until then I'd prefer not to merge failing tests to master (or release).

Sorry about that

Milacus commented:
Understandable. Could you please rerun the tests on the master? 

Either my test environment has some issues or the muster is currently not passing the tests either. I'll try to take a look on the weekend.

Modelmat commented:
Can't seem to figure out how to rerun it through GitHub actions.

My guess would probably be something has broken on master, maybe sphinx upgraded? Not sure exactly.

Milacus commented:
I finally came around to take a look at it...

First I updated the drawio version as the old ones are no longer available on GitHub. This caused me to change the default drawio path (changed for a while now in the drawio.deb package) and only use the old path as a fall back (could not try it with the old versions though).

The failing test cases can be narrowed down due to changes of drawio and sphinx in their latest versions compared to the expected result in the test cases.

**Test cases page-index and figure:**
- Dimensions are off (often only by one pixel)
- You reported the ticket https://github.com/jgraph/drawio-desktop/issues/254 which was "fixed" by adjusting their test cases in the came way as I had to do

**Test case figure:**
- Seems as if Sphinx does generate something else in the latest version than what's specified in the test case. Expected is a div with class == "figure" but its actually a figure tag that gets generated in the HTML

**Test case imgconverter:**
- From my understanding Sphinx will try to choose the "best" output of the images. Which does seem to be SVG for images displayed on webpages (which makes sens to keep them as vector graphics rather than converting them to PNG/ raster graphics). 


Please have a look at my elaborations and changes. Please, correct me if I'm wrong or you disagree.

Modelmat commented on tests/test_compat.py:
```diff
@@ -12,6 +12,6 @@ def test_imgconverter(directives: List[Tag]):
     # SVG to PNG by sphinx.ext.imgconverter
-    assert img["src"] == "_images/box.png"
-    assert img["alt"] == "_images/box.png"
+    assert img["src"] == "_images/box.svg"
```
This should be converted to `png` by `imgconverter`, this is the point of the test.

I'm not sure why this does not work.

Could you split this into another commit and instead disable the test rather than change it? I can create an issue to figure out what went wrong later.

Milacus replied:
Done :heavy_check_mark:

Modelmat commented on tests/test_options.py:
```diff
@@ -68,14 +68,14 @@ def test_width_height(images: List[Path]):
     # https://github.com/jgraph/drawio-desktop/issues/254
```
Could you remove this comment and mention this  URL in the commit message body? This is what has been fixed so we are adjusting the tests.

Milacus replied:
Done :heavy_check_mark:

Modelmat commented on tests/test_options.py:
```diff
@@ -124,9 +124,8 @@ def test_figure(content: Sphinx, directives: List[Tag]):
-        assert div.name == "div"
-        assert "figure" in div["class"]
+        imageContainerTag = img.parent
+        assert imageContainerTag.name == "figure"
```
Erm, this should still be a div? Not sure why this has changed?

Modelmat reviewed (changes requested):
Please split as requested and make the commit message clear ("Fix tests by adjusting the expected dimensions\n <url of drawio issue") or something similar.

Modelmat commented on .github/workflows/workflow.yml:
```diff
@@ -44,7 +44,7 @@ jobs:
           pip install -r dev-requirements.txt
       - name: Intall draw.io dependencies
         run: |
-          wget -O drawio.deb https://github.com/jgraph/drawio-desktop/releases/download/v12.6.5/draw.io-amd64-12.6.5.deb
```
This update is logically part of the adjustment for tests as this would fail the tests, and the new tests would fail without it. Please merge this commit & the commit for updating the tests.

Correction to my previous commit message suggestion:

```
Update draw.io package version

Also updates the expected dimensions for the tests as <url> was resolved.
```

Milacus replied:
Done :heavy_check_mark:

Modelmat commented on sphinxcontrib/drawio/deprecated.py:
```diff
@@ -166,7 +166,9 @@ def render_drawio(
     elif platform.system() == "Windows":
         binary_path = r"C:\Program Files\draw.io\draw.io.exe"
     else:
-        binary_path = "/opt/draw.io/drawio"
```
Please don't update the deprecated file. It should remain the same.

Milacus replied:
This means we'll have to change a bunch of the test cases and remove some in total (see the last commit)

Modelmat commented on tests/conftest.py:
```diff
@@ -97,14 +97,14 @@ def images(content: Sphinx) -> List[Path]:
 @pytest.fixture()
 def legacy_tex_images(content: Sphinx) -> List[Path]:
-    tex = (content.outdir / "python.tex").text()
+    tex = (content.outdir / "python.tex").read_text()
```
This should also be updated with the draw.io update.

Milacus replied:
Done :heavy_check_mark:

Milacus commented:
Apparently we do have very different development styles but I tried to change my commits according to your suggestions.

We need to create a new issue for the `'imgconverter'` test case, which is skipped for now.

When not adapting the path for the deprecated 'drawio' directive, a whole bunch of test cases fail (due to not finding the draw.io binary). I removed the test steps and test cases. Meaning, that the deprecated directive can also only be used with older versions of the draw.io package. We might need to emphasize this rather breaking change a bit more. What do you think?

Modelmat commented:
Uh, sorry about that. I didn't do most of the working creating the deprecated version so I had forgotten that there were tests for the deprecated version.

I'm okay with changing the path if it is necessary for the tests to fail (I'd rather keep the deprecated version working if possible). If you'd like to undo 58e6296 and just put the fix for the path into the deprecated file that'd be great :)

Also if you can just do a once over your commits for spelling, and I'll be happy to merge this.

Modelmat commented:
I think you have some commits with different authors (Milicaus vs josh?) Might want to fix that.

Also if you could merge the "fixed review" commit back into the first commit for fixing the preview.

Milacus commented:
Added the changes in the deprecated file and reverted the removal of test cases/ test steps :heavy_check_mark: 

I'm sorry for the bad English in the previous commit message. Should be fixed now.

Thank you for taking the time, being so patient and responding so quickly.

Modelmat commented:
You're welcome :)

Just the last couple of things I mentioned just before: https://github.com/Modelmat/sphinxcontrib-drawio/pull/62#issuecomment-901026260

And then I need to remember how to publish a new release to pypi for you :)

Milacus commented:
I was working on multiple machines which are apparently set up differently. Now everything should be addressed

Modelmat commented:
... And it's released. Thanks for the contribution!

Milacus commented:
Thank you very much. Tested it on my Linux and Windows machine and both are working great :+1: 

Thank you for the great extension
